Mixed-integer nonlinear decomposition toolbox in Pyomo (MindtPy)

bernalde opened the pull request:
## Summary/Motivation:
This pull request is to include the Mixed-Integer Decomposition Toolbox in Pyomo (MindtPy) as part of the contrib packages.

## Changes proposed in this PR:
- Adding the MindtPy solver which includes Outer Approximation algorithm for MINLP.
- To come in future PRs: Partial Surrogate Cuts, Generalized Benders Decomposition, and Extended Cutting Plane 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

bernalde committed:
Initial clean import from previous branch

bernalde committed:
Subsolvers options passed down

bernalde committed:
modifications back working, ready for Felicity pull

bernalde committed:
changed parameter from decompostion_strategy to the previous strategy

bernalde committed:
Starting update of the code

bernalde committed:
Changes on the new version still not working

qtothec committed:
Merge remote-tracking branch 'upstream/master' into mindtpy

qtothec committed:
Bugfixes and renaming for moved imports

qtothec commented:
I opened #594 to help resolve one of the errors I ran into. You might also try reading through http://pyomo.readthedocs.io/en/latest/developer_reference/expressions/design.html#expression-classes, but I still couldn't quite tell how to go about doing checking for SumExpression, or if there were hidden "gotchas" involved.

The others I saw are related to incorrect/too many/too few arguments to function calls. You should be able to address those, but let me know if you run into one that doesn't make any sense.

bernalde committed:
Reproduced error with current implementation for Qi

bernalde committed:
Qi asked me to commit

qtothec committed:
Starting to rearrange things

qtothec committed:
Renaming and modifying plugin behavior

qtothec committed:
Split massive file into separate modules

qtothec committed:
Continue moving code into separate modules

qtothec committed:
Various monkey patches to make things work. PSC and GBD still need to be fixed.

qtothec committed:
Remove unused LP subproblem code

qtothec committed:
Continue splitting out code into modules

qtothec committed:
Intermediate stable step for OA

bernalde committed:
Adding slight modification to testcase2

qtothec committed:
Merge branch 'gdpopt_dev' into mindtpy

qtothec committed:
Rename a file, various code cleanup

qtothec committed:
No need for reverting transformations. We are working on a model clone.

qtothec committed:
Do not directly import a model

qtothec committed:
Code consolidation. Renaming mip to mip_solve

qtothec committed:
Code consolidation + aligning naming.
I'm open to a debate about what the proper naming should be for some of 
the configuration options, but they should be the consistent across the 
two packages, because duplicated code isn't a good thing.

qtothec committed:
Merge branch 'gdpopt_dev' into mindtpy

qtothec committed:
As it turns out, I liked your naming better.

qtothec committed:
Further code consolidation

qtothec committed:
Merge remote-tracking branch 'upstream/master' into mindtpy

qtothec committed:
Updates in order to support import path changes in Pyomo master

Commit:
trying windows installation

Commit:
adding a couple of modified examples

qtothec committed:
Merge remote-tracking branch 'upstream/master' into mindtpy

qtothec committed:
Bringing MindtPy up to date with upstream Pyomo changes

qtothec committed:
Merge GDPopt changes from gdpopt_update branch

qtothec committed:
Merge branch 'gdpopt_update' into mindtpy

qtothec committed:
Updates to GDPopt to generalize utilities for MindtPy

qtothec committed:
[unstable] Intermediate working commit

qtothec committed:
Merge branch 'gdpopt_update' into mindtpy
Merging more recent changes to GDPopt into MindtPy

qtothec committed:
GDPopt updates to support MindtPy development.
Changed best_solution_found to store model rather than just list of variables.

qtothec committed:
Updates to MindtPy to simplify the codebase and bring it closer to a merge point.

qtothec committed:
Continued consolidation. Changing print() to logger.

qtothec committed:
Merge branch 'gdpopt_update' into mindtpy

qtothec committed:
Merge branch 'master' into mindtpy

qtothec committed:
Merge branch 'master' into mindtpy

qtothec committed:
Create Mindtpy stub

bernalde committed:
Documentation completed and typos fixed

qtothec committed:
slight fixes to docs

qtothec committed:
removing functionality that is not yet fully implemented

qtothec commented:
@bernalde We are running into this problem. Is this a nonlinear objective issue?

```
======================================================================
ERROR: Test the outer approximation decomposition algorithm.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/Pyomo/pyomo/pyomo/contrib/mindtpy/tests/test_mindtpy.py", line 100, in test_OA_MINLP_simple
    nlp_solver=required_solvers[0])
  File "/home/travis/build/Pyomo/pyomo/pyomo/contrib/mindtpy/MindtPy.py", line 342, in solve
    MindtPy_iteration_loop(solve_data, config)
  File "/home/travis/build/Pyomo/pyomo/pyomo/contrib/mindtpy/iterate.py", line 24, in MindtPy_iteration_loop
    solve_OA_master(solve_data, config)
  File "/home/travis/build/Pyomo/pyomo/pyomo/contrib/mindtpy/mip_solve.py", line 146, in solve_OA_master
    (master_terminate_cond, results.solver.message))
ValueError: MindtPy unable to handle MILP master termination condition of unbounded. Solver message: <undefined>
```

bernalde committed:
Corrected tests for Mindtpy with infeasible MILP master problem

qtothec committed:
Merge branch 'master' into mindtpy

qtothec commented:
I think that this is now ready for review

jsiirola commented on pyomo/contrib/mindtpy/tests/alan.py:
```diff
@@ -0,0 +1,49 @@
+#  MINLP written by GAMS Convert at 11/10/17 15:35:19
```
Please add a reference to the original model.  Since it is a direct translation of another model written in GAMS, we need to know the license under which the original model was released.

bernalde replied:
Solved in the last commit.

jsiirola commented on pyomo/contrib/mindtpy/tests/batchdes.py:
```diff
@@ -0,0 +1,84 @@
+#  MINLP written by GAMS Convert at 11/10/17 15:35:19
```
Please add a reference to the original model.  Since it is a direct translation of another model written in GAMS, we need to know the license under which the original model was released.

bernalde replied:
Solved in the last commit.

jsiirola commented on pyomo/contrib/mindtpy/tests/flay03m.py:
```diff
@@ -0,0 +1,100 @@
+#  MINLP written by GAMS Convert at 11/10/17 15:35:21
```
Please add a reference to the original model.  Since it is a direct translation of another model written in GAMS, we need to know the license under which the original model was released.

bernalde replied:
Solved in the last commit.

jsiirola reviewed (changes requested):
Overall, this looks good.  We need to address the licensing of some of the examples and then I think we will be good to go.

bernalde committed:
Adding references to MINLPLib problems

qtothec commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -0,0 +1,99 @@
+        if constr.body.polynomial_degree() in (1, 0):
+            continue
+        rhs = ((0 if constr.upper is None else constr.upper) +
+               (0 if constr.lower is None else constr.lower))
```
You may want to consider the case of `constr.upper = constr.lower = some number`.

bernalde replied:
After some thought, I have decided to subtract the lower bound if both exist. In case of an equality constraint, we will result in an RHS equal to the original RHS and in the next lines relaxed to be generated in whichever direction the dual says. If it has both bounds and are different, it uses as the RHS the upper bound, which will be looser and therefore less risky to cut stuff that we wouldn't like to cut (UB>=LB).
Solved in the last commit.

qtothec commented on pyomo/contrib/mindtpy/tests/eight_process_problem.py:
```diff
@@ -0,0 +1,149 @@
+                           Objective, Param, RangeSet, Var, exp, minimize)
+
+
+class EightProcessFlowsheet(ConcreteModel):
```
Should this be replaced by just importing the GDP model and calling `bigM` on it, or is it worth keeping around the MINLP reformulation separately?

bernalde replied:
I would go for the independent files for the moment, but I'm open to just use the Big-M transformation of the GDP.

qtothec reviewed (changes requested):
Glancing over things again. Noticed something we ought to address, and have a question.

bernalde committed:
Solved issue with cuts for constraints with both lower and upper bounds

qtothec committed:
Update cut_generation.py

Clean up implementation.
Add comment to explain code logic.

qtothec committed:
Merge remote-tracking branch 'upstream/master' into mindtpy

qtothec reviewed (approved):
Looks good, assuming that I didn't mess up and the tests pass.
